src/l500/l500-options.cpp

```diff
@@ -255,75 +380,207 @@ namespace librealsense
+                gain_and_laser_to_preset = {
+                    { { RS2_DIGITAL_GAIN_HIGH, def_laser }, RS2_L500_VISUAL_PRESET_NO_AMBIENT },
+                    { { RS2_DIGITAL_GAIN_LOW, max_laser }, RS2_L500_VISUAL_PRESET_LOW_AMBIENT },
+                    { { RS2_DIGITAL_GAIN_HIGH, max_laser }, RS2_L500_VISUAL_PRESET_MAX_RANGE },
+                    { { RS2_DIGITAL_GAIN_LOW, def_laser }, RS2_L500_VISUAL_PRESET_SHORT_RANGE },
+                };
+
+            auto gain = ( rs2_digital_gain )(int)_digital_gain->query();
+            auto laser = _hw_options[RS2_OPTION_LASER_POWER]->query();
+
+            auto it = gain_and_laser_to_preset.find( { gain, laser } );
+
+            if( it != gain_and_laser_to_preset.end() )
+            {
+                return it->second;
+            }
+            
+            return RS2_L500_VISUAL_PRESET_CUSTOM;
         }
-        else
+        catch( ... )
+        {
+            LOG_ERROR( "Exception caught in calc_preset_from_controls" );
+        }
+        return RS2_L500_VISUAL_PRESET_CUSTOM;
+    }
+
+    l500_preset_option::l500_preset_option(option_range range, std::string description, l500_options * owner)
+        :float_option_with_description< rs2_l500_visual_preset >(range, description),
+        _owner( owner )
+    {}
+
+    void l500_preset_option::set( float value )
+    {
+        if( static_cast< rs2_l500_visual_preset >( int( value ) )
+            == RS2_L500_VISUAL_PRESET_DEFAULT )
+            throw invalid_value_exception( to_string()
+                                           << "RS2_L500_VISUAL_PRESET_DEFAULT was deprecated! " );
+
+        _owner->verify_max_usable_range_restrictions( RS2_OPTION_VISUAL_PRESET, value );
+        _owner->change_preset( static_cast< rs2_l500_visual_preset >( int( value ) ) );
+
+        float_option_with_description< rs2_l500_visual_preset >::set( value );
+    }
+
+    void l500_preset_option::set_value( float value )
+    {
+        float_option_with_description< rs2_l500_visual_preset >::set( value );
+    }
+
+    void l500_options::on_set_option( rs2_option opt, float value )
+    {
+        auto advanced_controls = get_advanced_controls();
+        if( std::find( advanced_controls.begin(), advanced_controls.end(), opt )
+            != advanced_controls.end() )
         {
-            auto advanced_controls = get_advanced_controls();
-            if (std::find(advanced_controls.begin(), advanced_controls.end(), opt) != advanced_controls.end())
+            verify_max_usable_range_restrictions( opt, value );
+
+            if (opt == RS2_OPTION_DIGITAL_GAIN)
             {
-                verify_max_usable_range_restrictions( opt, value );
-                move_to_custom();
+                // WA for fw bug will be removed after fixed on FW
+                //std::this_thread::sleep_for( std::chrono::milliseconds( 500 ) );
+                update_defaults();
```

Comment:
Shouldn't we move this into digital_gain_option::set()?

Reply:
Let's move to the observer on digital_gain

Reply:
I moved the code handle gain to digital_gain_option class

Reply:
Then why is it still here??

Reply:
fixed